### PR TITLE
[1.4] Time dilation + more precise time control

### PR DESCRIPTION
**Pull request by ThomasThePencil**
### Overview
This PR aims to make it easier to more precisely control the flow of time, as well as finally making it possible for modders to force time, tile updates, or event updates (more on that in a second) to flow slower than normal.

### What all is done
- Added `Main.desiredWorldEventsUpdateRate`.
  - A lot of things are now tied to this instead of `Main.dayRate`, such as fallen star spawns, weather updates, and Travelling Merchant spawns.
  - By default, this is the same as `Main.dayRate`, mimicking vanilla behavior.
- All time-related fields have been made doubles with very few exceptions; the way a lot of things check for the passage of time has been adjusted to account for the new changes.
  - This means it is now possible to have `Main.dayRate` be lower than 1, causing time to flow slower than normal. The same applies for `Main.desiredWorldTilesUpdateRate` and the new `Main.desiredWorldEventsUpdateRate` field.

#### Why use doubles instead of floats?
Two reasons.
- Firstly and most importantly, this allows an incredible degree of precision when working with time, the rate at which it flows, and related fields --- which, given the integral importance of time to the flow of the game, is very important.
- Secondly and perhaps less importantly, `Main.time`, the field responsible for controlling the flow of time is a double --- it's nice to be consistent.

### CODE REVIEW

**Review by Mirsario — APPROVED** [review]

File: patches/tModLoader/Terraria/GameContent/Events/MysticLogFairiesEvent.cs.patch [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+ 				if (_delayUntilNextAttempt == 0) {
+ 					_delayUntilNextAttempt = 60;
+ 					TrySpawningFairies();
+@@ -60,12 +_,14 @@
+ 
+ 			int oneOverSpawnChance = GetOneOverSpawnChance();
+ 			bool flag = false;
++			_timePass += Main.desiredWorldEventsUpdateRate;
+-			for (int i = 0; i < Main.dayRate; i++) {
++			for (int i = 1; i < (int)_timePass; i++) {
```
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/GameContent/Events/MysticLogFairiesEvent.cs.patch:43]: This needs to be `<=`

File: patches/tModLoader/Terraria/GameContent/Events/Sandstorm.cs.patch [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+ 
+ 					if (Main.windSpeedCurrent == 0f)
+ 						TimeLeft = 0;
+@@ -41,10 +_,12 @@
+ 					int num = 21600;
+ 					num = ((!Main.hardMode) ? (num * 3) : (num * 2));
+ 					if (HasSufficientWind()) {
++						_timePass += Main.desiredWorldEventsUpdateRate;
+-						for (int i = 0; i < Main.dayRate; i++) {
++						for (int i = 1; i < (int)_timePass; i++) {
```
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/GameContent/Events/Sandstorm.cs.patch:40]: Also here

File: patches/tModLoader/Terraria/Graphics/Effects/SkyManager.cs.patch [hunk 3]
```diff
@@ -1,15 +1,52 @@
...
+ 		public void Update(GameTime gameTime) {
++			/*
+ 			int num = Main.dayRate;
+ 			if (num < 1)
+ 				num = 1;
++			*/
+ 
++			_timePass += Main.desiredWorldEventsUpdateRate;
+-			for (int i = 0; i < num; i++) {
++			for (int i = 1; i < (int)_timePass; i++) {
```
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/Graphics/Effects/SkyManager.cs.patch:27]: and here

File: patches/tModLoader/Terraria/Main.cs.patch [hunk 4]
```diff
@@ -1113,6 +1145,30 @@
...
+ 			if (cloudBGActive > 0f) {
+ 				if (cloudBGActive > 1f)
+-					cloudBGActive -= (float)dayRate / num3;
++					cloudBGActive -= (float)desiredWorldEventsUpdateRate / num3;
+ 
+ 				if (cloudBGActive < 1f)
+ 					cloudBGActive = 1f;
+ 
+-				if (cloudBGActive == 1f && rand.Next((int)((float)(num2 * 2 / Math.Max(dayRate, 1)) * num3)) == 0) {
++				if (cloudBGActive == 1f && rand.Next((int)((float)(num2 * 2 / Math.Max(desiredWorldEventsUpdateRate, 1)) * num3)) == 0) {
```
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/Main.cs.patch:1159]: These int casts seem like they'll round to 0 in some cases. Should use a float based rand.Next

File: patches/tModLoader/Terraria/Main.cs.patch [hunk 5]
```diff
@@ -1173,6 +1229,24 @@
...
+@@ -12671,10 +_,16 @@
+ 			if (netMode != 1)
+ 				updateCloudLayer();
+ 
++			if (desiredWorldEventsUpdateRate <= 0)
++				goto skipWeatherUpdates;
++
++			timePass += Main.desiredWorldEventsUpdateRate;
+-			for (int i = 0; i < dayRate; i++) {
++			for (int i = 1; i < (int)timePass; i++) {
```
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/Main.cs.patch:1241]: `<=`

File: patches/tModLoader/Terraria/Main.cs.patch [hunk 6]
```diff
@@ -3951,6 +4105,89 @@
...
+-							rainTime -= dayRate;
++							rainTime -= desiredWorldEventsUpdateRate;
+-							if (dayRate > 0) {
++							if (desiredWorldEventsUpdateRate > 0) {
+-								int num = 86400 / dayRate / 24;
++								double num = 86400 / desiredWorldEventsUpdateRate / 24;
+ 								if (rainTime <= 0)
+ 									StopRain();
+-								else if (rand.Next(num * 2) == 0)
++								else if (rand.Next((int)(num * 2)) == 0)
```
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/Main.cs.patch:4145]: random and int casts

File: patches/tModLoader/Terraria/ModLoader/IO/WorldIO.cs [hunk 7]
```diff
@@ -32,7 +33,8 @@ internal static class WorldIO
 				["bestiaryChats"] = SaveNPCBestiaryChats(),
 				["anglerQuest"] = SaveAnglerQuest(),
 				["townManager"] = SaveTownManager(),
-				["modData"] = SaveModData()
+				["modData"] = SaveModData(),
+				["alteredVanillFields"] = SaveAlteredVanillaFields()
```
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/ModLoader/IO/WorldIO.cs:37]: typo

File: patches/tModLoader/Terraria/ModLoader/IO/WorldIO.cs [hunk 8]
```diff
@@ -80,6 +82,7 @@ internal static class WorldIO
 				customDataFail = e;
 				throw;
 			}
+			LoadAlteredVanillaFields(tag.GetCompound("alteredVanilla"));
```
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/ModLoader/IO/WorldIO.cs:85]: key mismatch with above

File: patches/tModLoader/Terraria/WorldGen.cs.patch [hunk 9]
```diff
@@ -1559,15 +1567,41 @@
...
  
- 			for (int l = 0; l < Main.dayRate; l++) {
+@@ -46212,8 +_,9 @@
+ 			if (Main.dayTime)
+ 				return;
+ 
+-			for (int l = 0; l < Main.dayRate; l++) {
 -				float num6 = Main.maxTilesX / 4200;
++			_timePass += Main.desiredWorldEventsUpdateRate;
++			for (int l = 1; l < (int)_timePass; l++) {
```
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/WorldGen.cs.patch:1586]: `<=`

**Review by Chicken-Bones — CHANGES_REQUESTED** [review]

File: patches/tModLoader/Terraria/Main.cs.patch [hunk 10]
```diff
@@ -1175,7 +1175,7 @@
  				cloudBGActive = 0f;
  
 -			if (cloudBGActive == 0f && rand.Next((int)((float)(num2 * 12 / ((dayRate == 0) ? 1 : dayRate)) / num3)) == 0) {
-+			if (cloudBGActive == 0f && rand.Next(((num2 * 12 / ((desiredWorldEventsUpdateRate == 0) ? 1 : desiredWorldEventsUpdateRate)) / num3)) == 0) {
++			if (cloudBGActive == 0f && rand.NextDouble() <= 1 / ((num2 * 12 / desiredWorldEventsUpdateRate) / num3)) {
```
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/Main.cs.patch:1178]: change
`1 / ((num2 * 12 / desiredWorldEventsUpdateRate) / num3)`
to
`num3 * desiredWorldEventsUpdateRate / (num2 * 12 )`

File: patches/tModLoader/Terraria/Main.cs.patch [hunk 11]
```diff
@@ -3664,15 +3664,17 @@
...
-@@ -44801,7 +_,7 @@
+@@ -44801,8 +_,8 @@
  						float num5 = rand.Next((int)num3, (int)num4);
  						num5 *= (1f - maxRaining + 1f) / 2f;
  						num5 *= (1f - windSpeedTarget + 1f) / 2f;
 -						num5 *= (float)dayRate;
+-						if (rand.Next((int)num5) == 0)
 +						num5 *= (float)desiredWorldEventsUpdateRate;
- 						if (rand.Next((int)num5) == 0)
++						if (rand.NextFloat() <= 1f / num5)
```
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/Main.cs.patch]: This code is weird. It looks like, in vanilla, the faster the day is going, the rarer lightning gets?
Unfortunately that means if someone sets `desiredWorldEventsUpdateRate` small enough here, then you'll get lightning every frame...

**Review by Chicken-Bones — CHANGES_REQUESTED** [review]

File: patches/tModLoader/Terraria/Main.cs.patch [hunk 12]
```diff
@@ -3671,7 +3671,7 @@
 -						num5 *= (float)dayRate;
 -						if (rand.Next((int)num5) == 0)
 +						num5 *= (float)desiredWorldEventsUpdateRate;
-+						if (rand.NextFloat() <= 1f / num5)
++						if (rand.NextFloat() <= 1f / Math.Max(num5, 2400))
```
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/Main.cs.patch:3710]: I checked the surrounding code. `UpdateWeather` is called once per day 'tick' and if the day is running super slow, your other changes will ensure the calculations are correct. You should actually do `num5 *= Math.Max(desiredWorldEventsUpdateRate, 1)`

**Pull request merged**